unified-runtime/source/adapters/level_zero/v2/command_buffer.cpp

```diff
@@ -70,6 +169,284 @@ ur_exp_command_buffer_handle_t_::~ur_exp_command_buffer_handle_t_() {
+    uint32_t numUpdateCommands,
+    const ur_exp_command_buffer_update_kernel_launch_desc_t *updateCommands) {
+  auto supportedFeatures =
+      device->ZeDeviceMutableCmdListsProperties->mutableCommandFlags;
+  logger::debug("Mutable features supported by device {}", supportedFeatures);
+  for (std::size_t i = 0; i < numUpdateCommands; i++) {
+    auto commandDesc = updateCommands[i];
+    auto command = static_cast<kernel_command_handle *>(commandDesc.hCommand);
+    UR_ASSERT(this == command->commandBuffer,
+              UR_RESULT_ERROR_INVALID_COMMAND_BUFFER_COMMAND_HANDLE_EXP);
+
+    UR_ASSERT(!commandDesc.hNewKernel ||
+                  (supportedFeatures &
+                   ZE_MUTABLE_COMMAND_EXP_FLAG_KERNEL_INSTRUCTION),
+              UR_RESULT_ERROR_UNSUPPORTED_FEATURE);
+    // Check if the provided new kernel is in the list of valid alternatives.
+    if (commandDesc.hNewKernel &&
+        !command->validKernelHandles.count(commandDesc.hNewKernel)) {
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+
+    if (commandDesc.newWorkDim != command->workDim &&
+        (!commandDesc.pNewGlobalWorkOffset ||
+         !commandDesc.pNewGlobalWorkSize)) {
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+
+    // Check if new global offset is provided.
+    size_t *newGlobalWorkOffset = commandDesc.pNewGlobalWorkOffset;
+    UR_ASSERT(
+        !newGlobalWorkOffset ||
+            (supportedFeatures & ZE_MUTABLE_COMMAND_EXP_FLAG_GLOBAL_OFFSET),
+        UR_RESULT_ERROR_UNSUPPORTED_FEATURE);
+    if (newGlobalWorkOffset) {
+      if (!context->getPlatform()->ZeDriverGlobalOffsetExtensionFound) {
+        logger::error("No global offset extension found on this driver");
+        return UR_RESULT_ERROR_INVALID_VALUE;
+      }
+    }
+
+    // Check if new group size is provided.
+    size_t *newLocalWorkSize = commandDesc.pNewLocalWorkSize;
+    UR_ASSERT(!newLocalWorkSize ||
+                  (supportedFeatures & ZE_MUTABLE_COMMAND_EXP_FLAG_GROUP_SIZE),
+              UR_RESULT_ERROR_UNSUPPORTED_FEATURE);
+
+    // Check if new global size is provided and we need to update group count.
+    size_t *newGlobalWorkSize = commandDesc.pNewGlobalWorkSize;
+    UR_ASSERT(!newGlobalWorkSize ||
+                  (supportedFeatures & ZE_MUTABLE_COMMAND_EXP_FLAG_GROUP_COUNT),
+              UR_RESULT_ERROR_UNSUPPORTED_FEATURE);
+    UR_ASSERT(!(newGlobalWorkSize && !newLocalWorkSize) ||
+                  (supportedFeatures & ZE_MUTABLE_COMMAND_EXP_FLAG_GROUP_SIZE),
+              UR_RESULT_ERROR_UNSUPPORTED_FEATURE);
+
+    UR_ASSERT(
+        (!commandDesc.numNewMemObjArgs && !commandDesc.numNewPointerArgs &&
+         !commandDesc.numNewValueArgs) ||
+            (supportedFeatures & ZE_MUTABLE_COMMAND_EXP_FLAG_KERNEL_ARGUMENTS),
+        UR_RESULT_ERROR_UNSUPPORTED_FEATURE);
+  }
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t ur_exp_command_buffer_handle_t_::applyUpdateCommands(
```

Comment:
could you please split this function? The global/local group size update, updating memory objects and some other logic could be extracted to make this more readable.